### PR TITLE
AwsS3 Multipart Uploads

### PR DESCRIPTION
**Pull request by MonsieurLanza**
This is a modification of AwsS3 adapter to allow multipart uploads when files are bigger than a certain size.

Size limits / part size are passed as options to the adapter and set to defaults, respectively 5Go and
5Mo. 5Go being the maximum size AWS S3 accepts in a single upload, and 5Mo the minimum part size.

Limitations : won't work with string content, you have to pass a resource handle otherwise write() will return false. We may use InMemoryStream to bypass this limitation ?

PHP7.1 fails on docker image, not sure why. May need some help on this.

### CODE REVIEW

File: src/Gaufrette/Adapter/AwsS3.php [hunk 1]
```diff
@@ -331,12 +366,107 @@ protected function computePath($key)
...
+    /**
+     * @param $key
+     *
+     * @return mixed
+     */
+    protected function initiateMultipartUpload($key)
+    {
+        $options = $this->getOptions($key);
+        $result = $this->service->createMultipartUpload($options);
+        return $result['UploadId'];
```
Comment (nicolasmure) [inline src/Gaufrette/Adapter/AwsS3.php]: Please add an empty line before this return statement :wink:

File: src/Gaufrette/Adapter/AwsS3.php [hunk 2]
```diff
@@ -166,14 +187,28 @@ public function write($key, $content)
...
+            $success = true;
+
+            if ($size >= $this->sizeLimit && $isResource) {
+                $success = $this->multipartUpload($key, $content);
+            } else if ($size <= $this->sizeLimit) {
+                $this->service->putObject($options);
+            } else {
+                /*
+                 * content is a string & too big to be uploaded in one shot
+                 * We may want to throw an exception here ?
```
Comment (nicolasmure) [inline src/Gaufrette/Adapter/AwsS3.php]: It would lead into a BC break IMO.

File: src/Gaufrette/Adapter/AwsS3.php [hunk 3]
```diff
@@ -331,12 +366,107 @@ protected function computePath($key)
...
+            $key,
+            [
+                'UploadId' => $uploadId,
+                'PartNumber' => $partNumber,
+                'Body' => fread($content, $this->partSize),
+            ]
+        );
+
+        $options = $this->getOptions($key, $options);
+        return $this->service->uploadPart($options);
```
Comment (nicolasmure) [inline src/Gaufrette/Adapter/AwsS3.php]: Same here : please insert an empty line before the return statement :wink:

File: src/Gaufrette/Adapter/AwsS3.php [hunk 4]
```diff
@@ -331,12 +366,107 @@ protected function computePath($key)
...
+     * @param $key
+     * @param $content
+     *
+     * @return bool
+     */
+    protected function multipartUpload($key, $content)
+    {
+        $uploadId = $this->initiateMultipartUpload($key);
+
+        $parts = array();
```
Comment (nicolasmure) [inline src/Gaufrette/Adapter/AwsS3.php]: Please use short array `[]` syntax :wink:

**Comment by MonsieurLanza** [comment]
Hope I did not forget anything :)

**Comment by nicolasmure** [comment]
Can you rebase on master (to consider the merge of #514 ), and update your test please ([note the changes](https://github.com/KnpLabs/Gaufrette/pull/514/files#diff-61b9bc32613b9bc34d9ea4c28e31f601R76))?

**Comment by MonsieurLanza** [comment]
Done.

**Comment by MonsieurLanza** [comment]
Fix : Use Amazon's size limit for string upload and option specific limit for resource handle to allow 5Go upload no matter what.
Added a test case to cover that.

**Comment by MonsieurLanza** [comment]
Hello, some news here ? 
We would like this PR to get rid of our custom repo. :)

**Comment by amcastror** [comment]
Hi, are there plans to merge this? I'm having troubles with this.

Thanks!

**Comment by jfouca** [comment]
@NiR- we would really appreciate your review on this since @nicolasmure approved it some time ago :)

thank you

**Comment by amodliszewski** [comment]
Hello, any idea when this would be merged? Love the concept, would release some workarounds form my project.

File: src/Gaufrette/Adapter/AwsS3.php [hunk 5]
```diff
@@ -13,6 +14,12 @@
  */
 class AwsS3 implements Adapter, MetadataSupporter, ListKeysAware, SizeCalculator, MimeTypeProvider
 {
+    /*
+     * Amazon S3 does not not allow uploads > 5Go
+     */
+    const MAX_CONTENT_SIZE = 5368709120;
+    const DEFAULT_PART_SIZE = 5242880;
```
Comment (Nyholm) [inline src/Gaufrette/Adapter/AwsS3.php:21]: Why this default part size? Do 64Mb or something that make sense.

**Review by Nyholm — COMMENTED** [review]
The current default values will limit uploads to 50GB. Sure it is larger than 5GB but still not big enough.

**Comment by Nyholm** [comment]
Btw. Version 0.10.0 will support AsyncAws and that switches to multipart uploads automatically. 

It will not have a memory foot print that is related to the part size.

**Pull request closed without merging**